File: internal/services/accounts/accounts.go
// Package accounts creates users with their personal space and manages
// profile changes.
package accounts

import (
	"database/sql"
	"errors"
	"strings"
	"time"

	"dashboard/internal/crypto"
	"dashboard/internal/db"
	"dashboard/internal/enums"
	"dashboard/internal/model"
	"dashboard/internal/repos/content"
	"dashboard/internal/repos/users"
	"dashboard/internal/services/access"
	"dashboard/internal/services/audit"
)

// MinPassword is the minimum accepted password length.
const MinPassword = 12

// ErrEmailTaken means the email is already registered.
var ErrEmailTaken = errors.New("accounts: email already registered")

// ErrPasswordTooShort means the password is below MinPassword.
var ErrPasswordTooShort = errors.New("accounts: password too short")

// ErrWrongPassword means the current password did not match.
var ErrWrongPassword = errors.New("accounts: wrong current password")

// CheckPasswordRules validates a new password before hashing it.
func CheckPasswordRules(password string) error {
	if len(password) < MinPassword {
		return ErrPasswordTooShort
	}
	return nil
}

// Profile is the user-facing account summary shown in settings.
type Profile struct {
	ID           int64
	Email        string
	Name         string
	Role         enums.InstanceRole
	Locale       enums.Locale
	ColorMode    enums.ColorMode
	ThemeID      *int64
	StartBoardID *int64
	SearchEngine string
	TOTPEnabled  bool
	HasPassword  bool
	OIDCLinked   bool
	Prefs        map[string]any
}

// Create inserts a user plus their personal space. The caller owns the
// transaction (q may be a *sql.Tx), so this composes into a larger unit of
// work such as first-admin setup.
func Create(q db.Queryer, email, name string, password *string, role enums.InstanceRole, locale enums.Locale, sub string) (*model.User, error) {
	email = strings.TrimSpace(email)
	existing, err := users.ByEmail(q, email)
	if err != nil {
		return nil, err
	}
	if existing != nil {
		return nil, ErrEmailTaken
	}

	if password != nil {
		if err := CheckPasswordRules(*password); err != nil {
			return nil, err
		}
	}

	displayName := strings.TrimSpace(name)
	if displayName == "" {
		displayName = email
	}
	var hash string
	if password != nil {
		h, err := crypto.HashPassword(*password)
		if err != nil {
			return nil, err
		}
		hash = h
	}

	user := &model.User{
		Email: email, Name: displayName, PasswordHash: hash, Role: role, IsActive: true,
		Locale: locale, ColorMode: enums.ColorAuto, OIDCSub: sub, CreatedAt: time.Now().UTC(),
	}
	if err := users.Add(q, user); err != nil {
		return nil, err
	}

	space := &model.Space{Kind: enums.SpacePersonal, Name: user.Name, OwnerUserID: &user.ID, Version: 1}
	if err := content.AddSpace(q, space); err != nil {
		return nil, err
	}
	return user, nil
}

// TeamAssignment is one entry of a seed/invite's team list.
type TeamAssignment struct {
	Team string
	Role enums.TeamRole
}

// JoinTeams adds userID to each named team, creating missing teams.
func JoinTeams(q db.Queryer, userID int64, assignments []TeamAssignment) error {
	for _, a := range assignments {
		team, err := users.TeamByName(q, a.Team)
		if err != nil {
			return err
		}
		if team == nil {
			team, err = users.AddTeam(q, a.Team)
			if err != nil {
				return err
			}
		}
		role := a.Role
		if role == "" {
			role = enums.TeamViewer
		}
		if err := users.SetMember(q, userID, team.ID, role); err != nil {
			return err
		}
	}
	return nil
}

// GetProfile returns one user's account profile.
func GetProfile(d *sql.DB, who *access.Principal) (*Profile, error) {
	var p *Profile
	err := db.WithTx(d, func(tx *sql.Tx) error {
		u, err := users.Get(tx, who.UserID)
		if err != nil {
			return err
		}
		if u == nil {
			return sql.ErrNoRows
		}
		p = &Profile{
			ID: u.ID, Email: u.Email, Name: u.Name, Role: u.Role, Locale: u.Locale,
			ColorMode: u.ColorMode, ThemeID: u.ThemeID, StartBoardID: u.StartBoardID,
			SearchEngine: u.SearchEngine, TOTPEnabled: u.TOTPEnabled,
			HasPassword: u.PasswordHash != "", OIDCLinked: u.OIDCSub != "", Prefs: orEmpty(u.Prefs),
		}
		return nil
	})
	return p, err
}

// ProfileChanges lists the mutable profile fields UpdateProfile may set.
type ProfileChanges struct {
	Name         *string
	Locale       *enums.Locale
	ColorMode    *enums.ColorMode
	ThemeID      **int64
	StartBoardID **int64
	SearchEngine *string
}

// UpdateProfile applies the given (non-nil) changes to a user's profile.
func UpdateProfile(d *sql.DB, who *access.Principal, changes ProfileChanges) error {
	return db.WithTx(d, func(tx *sql.Tx) error {
		u, err := users.Get(tx, who.UserID)
		if err != nil {
			return err
		}
		if u == nil {
			return sql.ErrNoRows
		}
		if changes.Name != nil {
			u.Name = *changes.Name
		}
		if changes.Locale != nil {
			u.Locale = *changes.Locale
		}
		if changes.ColorMode != nil {
			u.ColorMode = *changes.ColorMode
		}
		if changes.ThemeID != nil {
			u.ThemeID = *changes.ThemeID
		}
		if changes.StartBoardID != nil {
			u.StartBoardID = *changes.StartBoardID
		}
		if changes.SearchEngine != nil {
			u.SearchEngine = *changes.SearchEngine
		}
		return users.Update(tx, u)
	})
}

// SetPref sets one key in a user's free-form preferences map.
func SetPref(d *sql.DB, who *access.Principal, key string, value any) error {
	return db.WithTx(d, func(tx *sql.Tx) error {
		u, err := users.Get(tx, who.UserID)
		if err != nil {
			return err
		}
		if u == nil {
			return sql.ErrNoRows
		}
		u.Prefs = orEmpty(u.Prefs)
		u.Prefs[key] = value
		return users.Update(tx, u)
	})
}

// ChangePassword verifies the current password (when one is set) and
// stores the new one.
func ChangePassword(d *sql.DB, who *access.Principal, current, newPassword, ip string) error {
	if err := CheckPasswordRules(newPassword); err != nil {
		return err
	}
	return db.WithTx(d, func(tx *sql.Tx) error {
		u, err := users.Get(tx, who.UserID)
		if err != nil {
			return err
		}
		if u == nil {
			return sql.ErrNoRows
		}
		if u.PasswordHash != "" && !crypto.CheckPassword(u.PasswordHash, current) {
			return ErrWrongPassword
		}
		hash, err := crypto.HashPassword(newPassword)
		if err != nil {
			return err
		}
		u.PasswordHash = hash
		if err := users.Update(tx, u); err != nil {
			return err
		}
		return audit.Log(tx, &who.UserID, "password.changed", "", ip, nil)
	})
}

func orEmpty(m map[string]any) map[string]any {
	if m == nil {
		return map[string]any{}
	}
	return m
}

File: app/services/accounts.py
"""Accounts: create users with their personal space, profile changes."""

from dataclasses import dataclass

from sqlalchemy.orm import Session

from app.db.base import session_scope
from app.db.models import Space, User
from app.enums import ColorMode, InstanceRole, Locale, SpaceKind, TeamRole
from app.repos import content, users
from app.services import audit, crypto, mail
from app.services import teams as team_service
from app.services.access import Principal

MIN_PASSWORD = 12


class AccountError(ValueError):
    pass


@dataclass
class Profile:
    id: int
    email: str
    name: str
    role: InstanceRole
    locale: Locale
    color_mode: ColorMode
    theme_id: int | None
    start_board_id: int | None
    search_engine: str | None
    totp_enabled: bool
    has_password: bool
    oidc_linked: bool
    prefs: dict


def check_password_rules(password: str) -> None:
    if len(password) < MIN_PASSWORD:
        raise AccountError("password.too_short")


def create(
    s: Session,
    email: str,
    name: str,
    password: str | None,
    role: InstanceRole = InstanceRole.USER,
    locale: Locale = Locale.DE,
    sub: str | None = None,
) -> User:
    """User plus personal space. Caller owns the transaction."""
    email = email.strip()
    if users.by_email(s, email):
        raise AccountError("account.email_taken")

    if password is not None:
        check_password_rules(password)

    user = users.add(
        s,
        User(
            email=email,
            name=name.strip() or email,
            password_hash=crypto.hash_password(password) if password else None,
            role=role,
            locale=locale,
            oidc_sub=sub,
        ),
    )
    content.add(s, Space(kind=SpaceKind.PERSONAL, name=user.name, owner_user_id=user.id))
    return user


def join_teams(s: Session, user_id: int, teams: list[dict]) -> None:
    """teams: [{"team": name, "role": "editor"}]; unknown teams are created."""
    for item in teams:
        team = users.team_by_name(s, item["team"]) or team_service.create_in(s, item["team"])
        users.set_member(s, user_id, team.id, TeamRole(item.get("role", TeamRole.VIEWER)))


def profile(who: Principal) -> Profile:
    with session_scope() as s:
        user = users.get(s, who.user_id)
        return Profile(
            id=user.id,
            email=user.email,
            name=user.name,
            role=InstanceRole(user.role),
            locale=Locale(user.locale),
            color_mode=ColorMode(user.color_mode),
            theme_id=user.theme_id,
            start_board_id=user.start_board_id,
            search_engine=user.search_engine,
            totp_enabled=user.totp_enabled,
            has_password=bool(user.password_hash),
            oidc_linked=bool(user.oidc_sub),
            prefs=dict(user.prefs or {}),
        )


def update_profile(who: Principal, **changes) -> None:
    allowed = {"name", "locale", "color_mode", "theme_id", "start_board_id", "search_engine"}
    with session_scope() as s:
        user = users.get(s, who.user_id)
        for key, value in changes.items():
            if key not in allowed:
                continue
            setattr(user, key, value)


def set_pref(who: Principal, key: str, value) -> None:
    with session_scope() as s:
        user = users.get(s, who.user_id)
        prefs = dict(user.prefs or {})
        prefs[key] = value
        user.prefs = prefs


def change_password(who: Principal, current: str | None, new: str, ip: str) -> None:
    check_password_rules(new)
    with session_scope() as s:
        user = users.get(s, who.user_id)
        if user.password_hash and not crypto.check_password(user.password_hash, current or ""):
            raise AccountError("password.wrong")

        user.password_hash = crypto.hash_password(new)
        audit.log(s, who.user_id, "password.changed", ip=ip)

    mail.security_notice(who.user_id, "password_changed")
